Unit Tests for Util

raavilagoo opened the pull request:
Unit Test for following files
1.Array
2.Bytes
3.Endian
4.Ranges
5. Vector

raavilagoo committed:
Unit Tests for Util

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,25 @@
+SCENARIO("Create an array", "[array]")
```
Since this file tests the `make_array` function, please use structure the SCENARIOs/GIVENs/WHENs/THENs with the standard pattern we are using for functions ([documented on Notion](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#51087c7b9af449d79163521b431a2e28)).

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,25 @@
+SCENARIO("Create an array", "[array]")
+  GIVEN("A set of int values '\x0f\x36\x08\xff\x89\x01\x80\x02' ")
+  WHEN("An array of given int size is created properly")
+     THEN("Value at index 5 in array is as expected")
```
Since we have knowledge about what the value at every index in the array should be, let's test every index, e.g. in a for loop comparing the returned array with a pre-constructed array containing the expected values in the expected order.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,25 @@
+
+  GIVEN("An array is given '\02\x03\x0a' ")
+  WHEN("Array is emptied")
+     THEN("The results are as expected")
```
We should also test for the correctness of `make_array` when we pass in objects by value, as well as when we pass in const references to objects, as well as when we pass in mutable references to objects, because part of the complexity of the implementation of `make_array` is that it uses `std::forward` to handle the latter two cases, and because we use that functionality when we make the `initializables` array in `main.cpp`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,14 @@
+SCENARIO("Get the byte from expected position", "[bytes]")
```
Since this file tests the `get_byte`, `set_byte<byte_index, T>`, and `set_byte<T>` functions, please use structure the SCENARIOs/GIVENs/WHENs/THENs with the standard pattern we are using for functions ([documented on Notion](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#51087c7b9af449d79163521b431a2e28)).

ethanjli replied:
We should test these functions with various data types (I would suggest `uint8_t`, `uint16_t`, and `uint32_t`) and various byte indexes (I would suggest using a for loop covering all possible byte indices for each data type, as well as indices which are out-of-range, e.g. a byte index of 5 for `uint32_t`).

ethanjli replied:
@raavilagoo It looks to me like the `set_byte` tests still need `WHEN`s where the byte indices are out-of-range for `uint16_t` and `uint32_t` outputs, not just for `uint8_t` outputs.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,14 @@
+     THEN("value is as expected")
+
+
+SCENARIO("Set byte ")
```
This scenario appears to test the `set_byte<byte_index, T>` function; we also need a scenario to test the `set_byte<T>` function (where byte_index is specified as a function parameter rather than a template parameter)

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Endian.txt:
```diff
@@ -0,0 +1,12 @@
+SCENARIO("read endian data")
```
Since this file tests the `read_bigend` and `write_hton` functions, please use structure the SCENARIOs/GIVENs/WHENs/THENs with the standard pattern we are using for functions ([documented on Notion](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#51087c7b9af449d79163521b431a2e28)).

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Endian.txt:
```diff
@@ -0,0 +1,12 @@
+SCENARIO("read endian data")
+  GIVEN("A buffer '\x01\x0f'")
+  WHEN("read buffer 2UL ")
+     THEN("reads correct value")
```
I would suggest stating the actual correct values, e.g. `0x010f`, to make clear how the bytes should be arranged in the outputs

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Endian.txt:
```diff
@@ -0,0 +1,12 @@
+
+  GIVEN("A buffer '\x02\x08\x00\x0d\x0e\x03\x0c\x00")
+  WHEN("read buffer 8UL ") 
+     THEN("reads correct value")
```
We also need tests for the `write_hton` functions.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,13 @@
+SCENARIO("Value is in provided range")
```
Since this file tests the `within` and `clamp` functions, please use structure the SCENARIOs/GIVENs/WHENs/THENs with the standard pattern we are using for functions ([documented on Notion](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#51087c7b9af449d79163521b431a2e28)).

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,13 @@
+SCENARIO("Value is in provided range")
+  GIVEN("A value '\x83'")
```
These tests should have coverage of all possible relationships between inputs and outputs:
- value < floor
- value == floor
- floor < value < ceiling
- value == ceiling
- value > ceiling

We should also have tests specifying what behavior we should expect when floor == ceiling and when floor > ceiling.

ethanjli replied:
@raavilagoo This applies not just to the `clamp` function, but also to the `within` function.

ethanjli replied:
Here's what I see about coverage on the `clamp` function:

Test cases needed (bolded items are covered)
1. **value < floor**
2. **value == floor**
3. **floor < value < ceiling**
4. value == ceiling
5. **value > ceiling** (this is covered twice)
6. floor == ceiling
  a. value < floor
  b. **floor == ceiling, value == floor == ceiling**
  c. floor == ceiling, value > ceiling
7. floor > ceiling
  a. value < ceiling
  b. floor > ceiling, value == floor
  c. floor > ceiling, value == ceiling
  d. **floor > ceiling, value > floor**

| test case | floor | test value | ceil |
| -- | -- | -- | -- |
| 3 | 1 | 131 | 255 |
| 5 | 1 | 131 | 10 |
| 1 | 10 | 8 | 131 |
| 2 | 8 | 8 | 131 |
| 5 | 8 | 83 | 15 |
| 6b | 8 | 8 | 8 |
| 7d | 15 | 83 | 8 |

ethanjli replied:
I would suggest grouping cases 1-5 in one scenario, then grouping cases 6a-c in another scenario, then grouping cases 7a-d in another scenario. This will also make it easier to verify test coverage of the cases.

ethanjli replied:
We also need 7e, floor > value > ceiling

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,22 @@
+SCENARIO( " Vectors are resized")
```
Since this file tests the `Vector` class, please use structure the SCENARIOs/GIVENs/WHENs/THENs with the standard patterns we are using for non-mutating methods and mutating methods ([documented on Notion](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#51087c7b9af449d79163521b431a2e28) as patterns 1 and 2).

ethanjli replied:
Tests should have coverage of all versions of all methods in the `Vector` class. In general, tests should cover when things (inputs and, if possible, internal state) are out-of-bounds (i.e. index > size), near bounds (i.e. index == size, index == size - 1, and index == size + 1), and within bounds (i.e. index < size - 1). Tests should have coverage of the behaviors of all methods when the vector has size 0, as well as when the vector is partially filled, as well as when the vector is completely full.

Tests need to be especially rigorous with the `erase` and `copy_from` methods, since that is where the logic is the most complicated and where hidden off-by-one errors, bounds-errors, and silent illegal accesses in the implementations are most likely to have occurred when I implemented those methods.

We also need tests confirming that the vector's state is correct when we directly modify the internal buffer using `Element *buffer()` and `resize` instead of using `push_back` and `erase`; this includes testing that the internal data should be valid after operations using those methods, and that the other methods (e.g. `push_back` and `erase`) work correctly after `Element *buffer()` and `resize` are used.

ethanjli replied:
For example, we need tests covering the behavior and state of the vector after `operator[]` is called with an out-of-bounds index (e.g. vector has capacity 20, is currently holding 10 elements, and then the `[]` operator is called for index 10, and for index 15).

ethanjli replied:
@raavilagoo For `Vector` methods which have something related to the internal state of a `Vector` object (which I believe is all methods in the `Vector` class), please use test [pattern 2](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#ffd035643205441ebb0ea3cdef473d8d) from the Notion document. In other words, the `GIVEN`s should be objects, not methods. Right now your txt file has the `GIVEN`s as methods. It looks like the cpp file has `GIVEN`s as objects, but I only look at the txt files for the `GIVEN`s/`WHEN`s/`THEN`s, and I always assume that the the txt files are the correct version of the tests while cpp files will eventually be changed to match the txt files.

ethanjli replied:
The approach to writing these tests should be to assume that each method of the object is or will become complicated and buggy, and tests should cover enough of the behavior of those methods that we can detect already-existing bugs and future regressions.

ethanjli replied:
We still need tests confirming that the vector's state is correct when we directly modify the internal buffer using `Element *buffer()` instead of using `push_back` and `erase`. Modifying data by using the `buffer` method should not modify the size of the `Vector` object, only the values inside the `Vector` at the indices which were touched using the `buffer` method.

ethanjli reviewed (commented):
Here's my review of how the tests are structured, and of additional test coverage needed.

raavilagoo committed:
Unit tests for Util

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,21 @@
+Scenario: The Util array function creates array correctly.
+  GIVEN("The Util make_array function")
+   WHEN("an array is created from 'n' uint8_t")
```
I don't know what `'n' uint8_t` means. I think it would be fine if you rephrased this to something like `WHEN('the function is called with 8 uint8_t values as the input parameters')`

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,21 @@
+Scenario: The Util array function creates array correctly.
+  GIVEN("The Util make_array function")
+   WHEN("an array is created from 'n' uint8_t")
+      THEN("the array is as expected")
```
This could be written as a more specific declaration, e.g. `THEN("The value of each element in the returned array matches the corresponding value given to the make_array function")` or even `THEN("The value at every index in the array is as expected")` (which is what you have in the cpp file but not the txt file)

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,21 @@
+
+
+  GIVEN("The Util make_array function ")
+   WHEN("an array is created from 'n' string")
```
This description makes me think that you are calling `make_array("n")`, which is clearly not what you do in the implementation of this test. I think it would be fine if you rephrased this to something like `WHEN('the function is called with 3 strings as the input parameters')`

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,21 @@
+
+  GIVEN("The Util make_array function") 
+  WHEN("an array is created from const references")
+    THEN("the array is as expected")
```
We should also confirm that when we change the values of the referenced variables, the values of the array will match those new values.

ethanjli replied:
@raavilagoo We still need to confirm that the array actually makes references rather than value copies, which we can test with the above approach

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,21 @@
+  GIVEN("The Util make_array function") 
+  WHEN("an array is created from const references")
+    THEN("the array is as expected")
+   
```
We are still missing a test for calling `make_array` with non-const references.

ethanjli replied:
In the test with non-const references, we should confirm that when we change the values of the referenced variables, the values of the array will match those new values. We should also confirm that when we modify the values of the array elements, values of the the referenced variables change accordingly.

ethanjli replied:
@raavilagoo We still need to confirm that the array actually makes references rather than value copies, which we can test with the above approach

ethanjli replied:
@raavilagoo

> In the test with non-const references, we should confirm that when we change the values of the referenced variables, the values of the array will match those new values.

I see the test for this.

> We should also confirm that when we modify the values of the array elements, values of the the referenced variables change accordingly.

I don't see the test for this yet

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Endian.txt:
```diff
@@ -0,0 +1,20 @@
+
+  GIVEN("Util function read_bigend")
+  WHEN("Util function read_bigend is used on  buffer 8UL ") 
+     THEN("Result is as expected")
```
Since these tests all are for different overloads of the same function, let's use the same pattern for the `WHEN`s and `THEN`s, i.e. the `THEN`s should all have the pattern `THEN('The function sets the result output parameter to 0x010f')`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Endian.txt:
```diff
@@ -0,0 +1,20 @@
+Scenario: Util function read_bigend works correctly
+  GIVEN("Util function read_bigend'")
+  WHEN("Util function read_bigend is used on  buffer 2UL ")
```
I think it would be better if you explicitly stated the contents of the buffer: `WHEN('The function is called with the buffer [0x01, 0x0f] as input data and a uint16_t result as the output parameter')`

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Endian.txt:
```diff
@@ -0,0 +1,20 @@
+    THEN("Result is as expected")   
+
+   WHEN("Util function write_hton is used on uint32_t")  
+     THEN("Result is as expected")
```
The `WHEN`s and `THEN`s for the `write_hton` tests should also list the data used: e.g. `WHEN('the function is called with the uint16_t value 0x0803 as input data and a two-byte buffer as the output parameter')` and `THEN('the function sets the output buffer to [0x08 0x03]')`. Make sure the implementation checks every byte in the output buffer, right now your implementations are only checking byte 0, and you may be causing `write_hton` to write into invalid memory locations because you only allocate a 1-byte output buffer when you need to provide 2-byte and 4-byte buffers for `uint16_t` and `uint32_t` inputs, respectively.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,31 @@
+Scenario: The Util Ranges  within function works correctly
+  GIVEN("The Util within function")
+  WHEN("floor is '\x01' and ceiling is '\0ff' , value is within range")
```
You should also explicitly state the value, e.g. `WHEN('The function is called with a floor of 0x01, a ceiling of 0xff, and a test value of 0x02')` or `WHEN('The function is called with a floor of 0x01, a ceiling of 0xff, and a test value of 0x01')` or `WHEN('The function is called with a floor of 0x01, a ceiling of 0xff, and a test value of 0x00')`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,31 @@
+
+Scenario:The Util Ranges  clamp function works correctly
+  GIVEN("The Util clamp function'")
+   WHEN("Value is equal to ceiling")
+      THEN("Returns correct value '\x83'")
+
+   WHEN("Value is greater than ceiling")
+     THEN("Returns expected value ceiling ")
+  
+  GIVEN("The Util clamp function ")
+  WHEN("Value is less than floor")
+    THEN("Returns expected floor value 0x0a")
+
+  WHEN("Value is equal to floor")
+    THEN("Returns expected  input value 0x08 ")  
+
+  GIVEN("The Util clamp function")  
+  WHEN("Value is in the range")
+    THEN("Returns expected  input value 0x53 ")
+
+
+  WHEN("floor is equal to ceiling")
+     THEN("Returns expected value ") 
+
+  WHEN("floor is greater than ceiling") 
```
All of these `WHEN`s should fully specify the values of the input parameters, e.g. `WHEN('the function is called with a floor of 0x00, a ceiling of 0x83, and a pre-clamped value of 0x83')`, or `WHEN('the function is called with a floor of 0xff, a ceiling of 0x83, and a pre-clamped value of 0x83')`. The `THEN`s should specify the return value, e.g. `THEN('the function returns 0x83 as the clamped value')`

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,31 @@
+
+
+  WHEN("floor is equal to ceiling")
+     THEN("Returns expected value ") 
```
What return value should we expect?

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,31 @@
+     THEN("Returns expected value ") 
+
+  WHEN("floor is greater than ceiling") 
+    THEN("Result is as expected ")   
```
What return value should we expect?

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+Scenario:The Util get_byte function correctly gets bytes from the given position
+   GIVEN("The util get_byte function")
+       WHEN("the get_byte function is called on 0x86 ")
+          THEN(""The value at the least significant byte position is "0x0a'")
```
Shouldn't the value at the LSB position be `0x86` rather than `0x0a`?

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+Scenario:The Util get_byte function correctly gets bytes from the given position
+   GIVEN("The util get_byte function")
+       WHEN("the get_byte function is called on 0x86 ")
+          THEN(""The value at the least significant byte position is "0x0a'")
+   GIVEN("The util get_byte function " )
+       WHEN("the get_byte function is called on 0x860a")  
+       THEN(""The value at the least significant byte position is "0x0a")
+       THEN("The value at the most signficant postiion is 0x86")
```
I would suggest specifying byte indices, rather than talking about MSB and LSB (because above uint16_t, there are bytes in between MSB and LSB). The `WHEN`s also need to specify the template parameters, as mentioned in our test patterns document on Notion: e.g. `WHEN('the function is called with byte index 0 and input type uint16_t as the template parameters and with 0x86 as the input value')` and `THEN('the returned value is 0x86')`, vs. `WHEN('the function is called with byte index 1 and input type uint16_t as the template parameters and with 0x86 as the input value')` and `THEN('the returned value is 0x00')`. Since the byte index specified as a function parameter, it should be specified in the `WHEN` and not the `THEN`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+
+   GIVEN("The util get_byte function'")
+       WHEN("the get_byte function is called on 0x86010f0a")
+       THEN("value is as expected ")
```
This `GIVEN` should have four `WHEN`-`THEN` pairs: the first `WHEN`+`THEN` is for byte index 0, the second `WHEN`+`THEN` is for byte index 1, etc.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+Scenario:The Util get_byte function correctly gets bytes from the given position
```
This test needs to specify `WHEN`+`THEN` pairs for calling `get_byte` with byte indices which are out-of-range.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+     THEN("Value is as expected 134 ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x86 for 0th index")
+     THEN("Value is as expected 0 ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x83 for 0th index")
+     THEN("Value is as expected in uint16_t data-type ")
+  
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x83 for 1st index")
+     THEN("Value is as expected in uint16_t data-type ")
+     
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x83 for 2nd index")
+     THEN("Value is as expected in uint32_t data-type ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x83 for 3rd index")
+     THEN("Value is as expected in uint32_t data-type ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x860a for 0th index")
+     THEN("Value is as expected in uint16_t data-type ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x860a for 1st index")
+     THEN("Value is as expected in uint16_t data-type ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x860a for 2nd index")
+     THEN("Value is as expected in uint16_t data-type ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x860a for 2nd index")
+     THEN("Value is as expected in uint32_t data-type ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x860a for 3rd index")
```
The `WHEN`s should specify when we're calling `set_byte` with byte_index specified as a template parameter vs. when we're calling it with byte_index specified as a regular input parameter. See also my comment on the `get_byte` tests about needing to specify all template parameters. So these `WHENs` should look like e.g. `WHEN('The function is called with byte index 1 and output type uint16_t as the template parameters and 0x83 as the input value')`

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+Scenario: The Util set_byte function correctly sets bytes to the given position
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x86 for 0th index")
+     THEN("Value is as expected 134 ")
```
Let's use hex for specifying the return values of all byte-level operations (i.e. everything in Bytes.txt), e.g. `THEN("The returned value is 0x86")`

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x83 for 0th index")
+     THEN("Value is as expected in uint16_t data-type ")
```
Let's specify the returned values for all `THEN`s, e.g. `THEN("The returned value is 0x0083")` or `THEN("The returned value is 0x8300")`or `THEN("The returned value is 0x0000")`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+     THEN("Value is as expected 134 ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x86 for 0th index")
```
Did you mean to specify that this is for byte_index 1, rather than byte_index 0? You already covered byte_index 0 in the previous `WHEN`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+     THEN("Value is as expected in uint16_t data-type ")
+
+  GIVEN("The Util set_byte function ")
+  WHEN("the get_byte function is called on 0x860a for 2nd index")
```
This `WHEN` look like a duplicate of the previous `WHEN`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,60 @@
+Scenario: The function in Util Vectors works correctly
```
Tests in this file are on the methods in the `Util::Vector` class, so the wording of the scenarios should match; you can also define scenarios by functionality, but it doesn't make sense to say "the function in Util push_back". Valid ways to write scenarios include: `Scenario: simple methods of the Vector class work correctly`, or `Scenario: insertion and deletion of elements in Vector at arbitrary indices works correctly`, or `Scenario: copying of elements to/from Vectors works correctly`

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,60 @@
+Scenario: The function in Util Vectors works correctly
+  GIVEN("The util resize function")
```
The `resize` method should be tested with the following possible conditions:
1. We try to resize a `Vector` beyond its maximum allowed size. This should fail with an error code. Afterwards, the `Vector` should behave as if we never tried to do a resize at all. The data in the `Vector` before the resize should not be corrupted. 
2. We try to resize a `Vector` to a larger length than its current length. The data in the `Vector` before the resize should not be corrupted.
3. We try to resize a `Vector` to a smaller length than its current length. The `Vector` should behave as if the data after the new size no longer exists, i.e. elements at those indices get overwritten by `push_back`, while attempts to interact with those indices using `erase` would fail with bounds errors.

In every condition, after the resize:
- The Vector should correctly report its size (what values are returned by `empty`, `full`, `size`, and `available`?).
- All indices which we expect to be out-of-bounds (e.g. for erase) should actually be out-of-bounds, while all indices which we expect to be in-bounds (e.g. for erase) should actually be in-bounds.

ethanjli replied:
@raavilagoo It's not clear to me that this has been fully covered (e.g. the push_back tests in condition 3), but let's save any remaining improvements for future PRs.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,60 @@
+    THEN("Result is as expected")
+
+    
+Scenario: The function in Util push_back works correctly
```
The `push_back` method should be tested with the following possible conditions:
- The `Vector` has more than one space available before `push_back`. `push_back` should succeed, and the next `push_back` should also succeed.
- The `Vector` has one space available before `push_back`. `push_back` should succeed, and the next `push_back` should fail with a bounds error.
- The `Vector` is completely full before `push_back`. `push_back` should fail with a bounds error. Afterwards, the `Vector` should behave as if the `push_back` method hadn't been called at all.

In every condition, after the `push_back`:

- The Vector should correctly report its size (what values are returned by `empty`, `full`, `size`, and `available`?).
- All indices which we expect to be out-of-bounds (e.g. for erase) should actually be out-of-bounds, while all indices which we expect to be in-bounds (e.g. for erase) should actually be in-bounds.
- All elements which were previously in the `Vector` should remain unchanged.
- If `push_back` succeeded, a new element should have been added to the end of the vector, and that element should have the value given as input to `push_back`, while every other element in the vector remains unchanged.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,60 @@
+
+
+  GIVEN("the Util push_back function ")
+  WHEN("the util push_back is called on string")
```
I don't understand what the purpose of this `WHEN` is.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,60 @@
+Scenario: The function in Util Vectors works correctly
+  GIVEN("The util resize function")
+  WHEN("the util resize function is called on 252UL buffer size")
+    THEN("Result is as expected ")
```
For non-`const` methods, we need to be more specific than "result is as expected". Specifically, we need to test:
- The value of the method's return
- The behavior of the `Vector` after the method is called, i.e. what happens when we call the same method or other methods afterwards.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,60 @@
+     THEN("Result is as expected")   
+
+Scenario:  Data is erased from particular index   
+   GIVEN("the Util erase function ")
```
The `erase` method should be tested with the following possible conditions:
- The `Vector` has more than one element before `erase`. If the `index` input parameter is in-bounds, `erase` should succeed, and the next `erase` (with an in-bounds input parameter) should also succeed. If the `index` input parameter was out-of-bounds for the first `erase` call, the `Vector` should behave as if the `erase` method hadn't been called at all.
- The `Vector` has exactly one element before `erase`. `erase(0)` should succeed (while `erase` with any other index should fail, and afterwards the `Vector` should behave as if the `erase` method hadn't been called at all), and the next `erase` should fail with a bounds error regardless of its input parameter.
- The `Vector` is empty before `erase`. `erase` should fail with a bounds error regardless of its input parameter. Afterwards, the `Vector` should behave as if the `erase` method hadn't been called at all.

In every condition, after the `erase`:

- The Vector should correctly report its size (what values are returned by `empty`, `full`, `size`, and `available`?).
- All indices which we expect to be out-of-bounds (e.g. for erase) should actually be out-of-bounds, while all indices which we expect to be in-bounds (e.g. for erase) should actually be in-bounds.
- All elements which were previously in the `Vector` should remain unchanged, except for the removal of the element at the specified index if the `erase` succeeded.

ethanjli replied:
@raavilagoo The tests in the scenario for the erase method do not seem to have THENs checking whether "All elements which were previously in the Vector should remain unchanged, except for the removal of the element at the specified index if the erase succeeded.". We really do need these checks; let's address them in a future PR.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt:
```diff
@@ -0,0 +1,60 @@
+     THEN("Result is as expceted")
+
+
+Scenario: The function in Util copy_from works correctly
```
We need tests for various possible combinations of values of the `dest_start_index` input parameter, sizes of the source collection, and sizes of the destination `Vector`.
- This includes `dest_start_index == 0`, non-zero `dest_start_index` which only overwrites elements in the vector, `dest_start_index == vector.size()` (so it only appends elements to the vector), and `dest_start_index` which overwrites some elements in the vector and appends some elements in the vector.
- This also includes testing inputs which are in-bounds (e.g. copying 5 elements into a vector with more than 5 spaces available), completely out-of-bounds (e.g. copying into a vector which is already completely full), or partially out-of-bounds (e.g. copying 5 elements into a vector with only 3 spaces available).
- Additionally, because I know that `copy_from` uses `memcpy`, we should make sure to test `copy_from` when `sizeof(Element) != 1` (e.g. if `Element` is a `uint16_t` rather than a `uint8_t`).

In every condition, after the `copy_from`:

- The Vector should correctly report its size (what values are returned by `empty`, `full`, `size`, and `available`?).
- All indices which we expect to be out-of-bounds (e.g. for any subsequent call to `erase`, and for any subsequent call to `copy_from`) should actually be out-of-bounds, while all indices which we expect to be in-bounds (e.g. for any subsequent call to `erase`, and for any subsequent call to `copy_from`) should actually be in-bounds.
- All elements which were previously in the `Vector` should remain unchanged, except for those within the range which `copy_from` was supposed to overwrite based on `dest_start_index`. All elements which were copied over should be in the correct order. For the `copy_from(const Element *source, ...)` overload variant, elements in `source` with index past `source_size` should not have been copied over. For the `copy_from(const Vector<...> source, ...)` overload variant, elements which were "discarded" after `resize` was called to decrease the size of the source `Vector` should not have been copied over.
- Copies made of elements from the source collection should be shallow copies, not deep copies.

Because we need so many test cases for `copy_from`, it might make sense to have one scenario for each of the three overloading variants of `copy_from`.

ethanjli replied:
@raavilagoo It's not clear to me that this has been fully covered (e.g. copy_from for elements which are not `uint8_t`, and preferably testing for struct elements), but let's save any remaining improvements for future PRs.

ethanjli reviewed (commented):
Here's my review of the txt files.

@rohanpurohit For future team members I think we need to develop an on-boarding guide for unit tests which explains in more detail our process for drafting the test specifications, e.g. txt vs. cpp files, test patterns, and coverage.

raavilagoo committed:
Unit Test for Util

raavilagoo commented:
Hi @ethanjli , I have updated this branch with expected changes.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Array.txt:
```diff
@@ -0,0 +1,29 @@
+Scenario: Empty an array
+  GIVEN('The make_array function')
+  WHEN('function empty is used on array (0x02, 0x03, 0x0a) ')
+     THEN('Result is as expected, array is empty')
```
`std::array::empty()` returns a bool representing whether the array is empty, but the array clearly isn't empty. In this case, the `empty()` method returns `false`, which means the array is empty. In the cpp file, you're testing `empty() == 0`, but actually it should be written as the equivalent statement `empty() == false` (which is more readable) and the `THEN` should declare that the array is **not** empty.

ethanjli replied:
@raavilagoo This issue has not yet been addressed.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,33 @@
+Scenario: The  Ranges:  within function works correctly
```
Since the `within` and `range` functions are primarily meant to be used on quantities/measurements rather than raw bytes, I think values in this file should be listed in base ten rather than hexadecimal. e.g. we should say `1` instead of `0x01`, `255` instead of `0xff`, `131` instead of `0x83`, etc.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,33 @@
+  GIVEN('The clamp function')
+   WHEN('the function is called with a floor of 0x01, a ceiling of 0xff, and a test value of 0x83')
+      THEN('Returns correct value 0x83')
+      THEN('Returns true ,  value 0x83 is in range')
```
The `clamp` function should return 131, it should not return `true`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,33 @@
+
+   WHEN('The function is called with a floor of 0x01, a ceiling of 0x0a, and a test value of 0x83')
+     THEN('Returns expected value 0x0a')
+     THEN('Returns true ,  value 0x0a is in range')
```
The `clamp` function should return 10, it should not return `true`.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,83 @@
+Scenario:The get_byte function in Bytes correctly gets bytes from the given position
+   GIVEN('get_byte function')
+       WHEN('the get_byte function is called for 0th index on 0x86 ')
+          THEN('The value at the 0th index  is 134')
```
Since the `get_byte` and `set_byte` functions operate on raw bytes rather than quantities/measurements, all values in this file should be reported in hexadecimal (e.g. `0x0a` instead of `10`, `0x00` instead of `0`, and `0x86` instead of `134`).

ethanjli replied:
@raavilagoo This issue has not been addressed for the THENs in this file.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,83 @@
+         THEN('Result is as expected, value at 3rd index is 134')
+
+      WHEN('the get_byte function is called for 4th index on 0x86010f0a')
+         THEN('Result is as expected, value at 4th index is 10')
```
Shouldn't the value at the 4th index be `0x00`? If this test is passing, I must be misunderstanding something, or maybe there's a problem in the implementation of `get_byte` vs. how I assumed it would behave. If this test is passing, it looks like the index wraps around past the end of the function.

ethanjli replied:
Filed as a bug in `get_byte` in #379

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,83 @@
+     THEN('Value is as expected  2197815296')
+
+  WHEN('the set_byte function is called on uint32_t data-type 0x83 for 4th index')
+     THEN('Value is as expected  131')
```
Hmm, I don't understand this test either. `131` is `0x83`, so apparently `set_byte<uint32_t, 4>(0x83)` returns `0x83`; I would've expected it to return `0x00`. If `set_byte` loops around when the index is too big, we'll need to include a warning about that in the documentation.

ethanjli replied:
Filed as a bug in `set_byte` in #379

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,83 @@
+Scenario:The get_byte function in Bytes correctly gets bytes from the given position
+   GIVEN('get_byte function')
+       WHEN('the get_byte function is called for 0th index on 0x86 ')
```
The WHENs in this file need to specify the template parameters. Refer to https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r646208849 (which I've marked as resolved because Github doesn't provide a way of replying to that comment)

ethanjli replied:
@raavilagoo This issue has not been addressed for every WHEN in this file.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,83 @@
+
+Scenario: The  set_byte function correctly sets bytes to the given position
+  GIVEN('The set_byte function')
+  WHEN('the set_byte function is called on 0x86 for 0th index')
```
The WHENs for `set_byte` should also fully specify the template parameters.

raavilagoo replied:
Hi @ethanjli , I have made required changes, this branch is ready for review.

ethanjli replied:
This issue has been addressed for the WHENs for `set_byte` on the function calls with `uint16_t` and `uint32_t` types, but not for the `set_byte` function calls with `uint8_t` types, i.e. you missed addressing this issue for the WHEN which this comment was added on and for the next WHEN (on line 46).

ethanjli reviewed (commented):
Here's my review of all the `txt` files except `vector.txt`; I've also left some previous comments on these files unresolved. The `vector` tests are complicated because we're testing a class with a lot of internal state, so let's discuss that in today's meeting.

raavilagoo committed:
Unit Tests for Util files
1. Added more tests in vector,
2. Ranges test in integer format
3.Added test in Array

raavilagoo committed:
Resolved conflicts with Develop branch

raavilagoo committed:
Merge Develop

raavilagoo committed:
Catch2-Clang-tidy fixes

raavilagoo committed:
Format fixes

raavilagoo committed:
Unit tests for Util
1.Bytes - Modified WHEN and THEN statements, added results for get_byte and set_byte in hex format
2.Array - Added const reference and non-const reference test cases
3.Ranges - Added more test cases

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-util-test

rohanpurohit committed:
vector: define structure for vector unit tests

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,33 @@
+
+  WHEN('an array is created from const references, value of reference variable is changed to -0x0c and 0x0a')
+    THEN('the array is as expected at every index [0x0c,0x0a]')
+    THEN('Returns correct changed values for reference variables [0x0c, 0x0a]')
```
I think it would be clearer to construct these tests like so:

- Given the make_array function
  - When an array is created of const references to two variables with values 0x02 and 0x0f
    - Then the array has the expected values, 0x02 and 0x0f
    - Then changing the value of the first referenced variable to 0x0c also changes the value at the first index of the array to 0x0c
    - (etc.)

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,33 @@
+    THEN('the array is as expected at every index [0x02,0x86]')
+    THEN('Returns correct changed values for reference variables [0x02, 0x86]')
+    
+Scenario: Empty an array
```
This scenario is not about making an array empty.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,33 @@
+    
+Scenario: Empty an array
+  GIVEN('The make_array function')
+  WHEN('function empty is used on array (0x02, 0x03, 0x0a) ')
```
How are we using the `make_array` function in this test? I don't see a reference to making an array using `make_array`. If you're doing what I think you're doing, perhaps you could just include "Then the empty method of the array returns false" as a THEN in each of the above WHENs.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+Scenario:The get_byte function in Bytes correctly gets bytes from the given position
+   GIVEN('get_byte function')
+       WHEN('the function is called with byte index 0 and input type uint8_t as the template parameters and with 0x86 as the input value ')
+          THEN('The value at the 0th index  is 0x86')
```
The tests in this file all look acceptable to me. If you want, you can make the THENs more concise by saying things like: THEN the function returns 0x86. Or, if you feel that the WHEN doesn't express the intent of the arguments clearly enough, you can say things like THEN the function returns the value at index 0, which is 0x86.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+       WHEN('the function is called with byte index 1 and input type uint8_t as the template parameters and with 0x86 as the input value')
+          THEN('The value at the 1st index  is 0x00')
+
+
```
What happens if we call `get_byte<2, uint8_t>(0x86010f0a)`? If this expression compiles, we should probably document this kind of behavior by defining unit tests for calling `get_byte` with input values which are wider than the types specified in the templated calls to `get_byte` (including tests for `uint8_t`, `uint16_t`, and `uint32_t`). It would make sense to create a new scenario for those tests.

If that expression doesn't compile, then we don't need to define unit tests for it.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+Scenario: The  set_byte function correctly sets bytes to the given position
+  GIVEN('The set_byte function')
+  WHEN('the function is called with byte index 0 and input type uint8_t as the template parameters and with 0x83 as the input value')
+     THEN('Value is as expected 0x86')
```
The THEN should specify that this value you're describing is the returned value: e.g. "THEN the returned value is as expected, 0x86".

The returned value in this test case, 0x86, is not what I would expect given the input value, 0x83.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,59 @@
+     THEN('Value is as expected 0x86')
+
+  WHEN('the function is called with byte index 1 and input type uint8_t as the template parameters and with 0x83 as the input value')
+     THEN('Value is as expected 0x00 ')
```
(note for future reference: after #379 is fixed, I expect to see that test cases like this one will fail to compile)

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.cpp:
```diff
@@ -0,0 +1,126 @@
+    WHEN(
+        "an array is created from const references, value of reference variable is changed to "
+        "-0x0c and 0x0a") {
+      expected = 0x0c;
+      array = 0x0a;
+      uint8_t const &ri = expected;
+      uint8_t const &pi = array;
+      auto data = PF::Util::Containers::make_array<uint8_t const>(ri, pi);
```
This is an incorrect implementation of the test because the values of the reference variables are changed before the array is made, rather than being changed after the array is made. The array should be created from `ri` and `pi`, and then _afterwards_ the values of `ri` and `pi` should be changed and the values in the array should be changed. You may need to use type `make_array<const std::reference_wrapper<uint8_t>>` or `make_array<std::reference_wrapper<const uint8_t>>` if you can't compile with type `make_array<uint8_t &>`.

ethanjli replied:
@rohanpurohit Since this is Raavi's first unit test PR, perhaps we also need to have review of the test implementations after each test description file is finalized. How much of this do you think it'd be reasonable for you to review, vs. how much of it should I review?

rohanpurohit replied:
@ethanjli yeah, I can definitely review them, maybe once I have reviewed them all you could take a brief overview and point out something that me or raavi missed?

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,100 @@
+Scenario:The Ranges:  clamp and within function works correctly
+  GIVEN('The clamp and within function')
+//floor>value>ceiling
```
What I see in your WHENs for this GIVEN is that mathematically `floor < value` and `value < ceiling`, i.e. `floor < value < ceiling`, not `floor > value > ceiling`. Some other comments in this file need to be corrected in this way, too.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,100 @@
+Scenario:The Ranges:  clamp and within function works correctly
+  GIVEN('The clamp and within function')
```
@rohanpurohit Last week I went over this file with you and discussed some other possible ways to rework the structure of Ranges (e.g. having the GIVENs be the floor, value, and ceiling, rather than GIVEN clamp and within functions). Would you be able to work with Raavi over Zoom to interactively rework the structure, as well as making other improvements along our overall test pattern (e.g. the WHENs in this file should name which function is being called, which is not happening consistently in this GIVEN)?

rohanpurohit replied:
Yes, @ethanjli I can do that. we'll go over several iterations and maybe you can review once we think it's ready for review from our end.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,100 @@
+  WHEN('The function within is called with a floor of 8, a ceiling of 255, and a test value of 83')  
+    THEN('Returns true, value is in range')
+
+//floor = ceiling = value
```
For example, this should clearly be `floor = ceiling < value`. Once these comments are corrected, then it'll be easier for me to confirm coverage of https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r654132679 .

ethanjli reviewed (commented):
Here's my review of the latest changes. There is still room for improvement with the correctness, quality and clarity of the test descriptions, but there's been good progress.

The kinds of comments I'd like to be making in reviews of unit test PRs are about higher-level concerns, like https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r655892364 (which is about how we can communicate the intent of tests), https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r655894098 (which is about a place where additional coverage would be helpful), https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r655906959 (which is about how we might split up one scenario into smaller scenarios), or https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r649635495 (which identified a way to improve the design of some code being tested, based on the actual behavior of the test). These kinds of things are more about building intuition for how to write test specifications, and I think these are skills which we can only develop by practicing together and giving/receiving feedback. If we can address issues like https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r655905893 or https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r655895167 (which are about errors that would be caught by proofreading, but might only be caught after getting good sleep and coming back with a fresh mind) or the part in https://github.com/pez-globo/pufferfish-software/pull/366#discussion_r655903975 about listing the name of the function being tested in each WHEN (which is something which should be explained in documentation, checklists, and/or training - so it's likely that we do not have adequate documentation for this) - before they reach formal PR review - then we should explore practices we could try to enable that. Here are some ideas I generated from brainstorming (probably only a small subset of these ideas would actually be helpful, but I was just brainstorming):
1. More strictly enforcing the rule to create, review, and iterate on the `txt` test specification files before making any `cpp` test implementation files; maybe even allowing a PR for test implementations only after an initial PR for the test specifications is reviewed and merged
2. Trying to do some live "pair programming" for creating the test specifications
3. Reserving time in our regular meetings to discuss the test specifications before proceeding to full PR review on Github
4. Creating more direct templates for test specifications - ones which could be copy-pasted
5. Creating checklists for what to check during self-review on a PR before requesting review from someone else - either for unit test PRs specifically, or for all PRs in general.
6. Scheduling separate meetings where the person who wrote the test specifications explains each test to someone else, before formal PR review on Github
7. Finding ways to document our approach to writing test specifications in more detail
8. Finding ways to document our approach to creating and reviewing PRs in more detail
9. Requiring two people to be coauthors of each test specification, i.e. they're both responsible for working together to put together the test specification for each test
10. When someone submits new code in a PR, requiring them to write the test specifications for at least some basic tests, and maybe also to write some explanation of what kinds of additional tests are needed (this could be part of the PR checklist), to act as a guide for writing future test specifications
11. Writing more documentation about what test specifications should look like, and including examples of specific test specifications to show the standard patterns to follow, and also discussing places where we might want to deviate from the standard patterns in order to make the tests easier to understand.

Since we are planning to expand the team, we will have a larger set of people writing different test specifications at different times and for different parts of the software - so let's use this as an opportunity to start a longer process of figuring out how we can work efficiently to ensure consistent quality and understandability across our unit tests and across people with different levels of familiarity in the practices we're trying to follow for writing unit tests. @renjipanicker @rohanpurohit @raavilagoo I think we'll need to have ongoing whole-team discussions for this, since it'll become more critical as the team expands, and since we'll probably only be able to figure it out by trial-and-error. Renji, do you have thoughts on how we can approach the process of figuring out what will work for us as a team, e.g. reserving time at regular meetings to discuss this, or doing separate meetings just for this topic, etc.?

raavilagoo committed:
Unit test for Util
1.Array
2.Bytes
3.Ranges
4.Vector

raavilagoo committed:
Merge remote-tracking branch 'raavi/feature/firmware-util-test' into feature/firmware-util-test

raavilagoo committed:
Modified Unit tests for Util

raavilagoo committed:
Bytes.h and Ranges.h changed for Issues 380 and 379

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Util/Bytes.h:
```diff
@@ -26,6 +26,7 @@ namespace Pufferfish::Util {
  */
 template <size_t byte_index, typename T>
 constexpr uint8_t get_byte(T number) noexcept {
+  static_assert(byte_index <= sizeof(T), "the byte index is greater than size of the type");  
```
Let's say `T = uint8_t`. Then it only has a single byte, which can be found at `byte_index = 0`. This means that we should not accept `get_byte<1, uint8_t>`, right? Currently this `static_assert` accepts that combination of template parameters.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Util/Ranges.h:
```diff
@@ -50,6 +50,9 @@ constexpr bool within(Numeric value, Numeric floor, Numeric ceiling) {
  */
```
#380 specifies that we need to fix not just the `clamp` function, but also the `within` function.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Util/Ranges.h:
```diff
@@ -50,6 +50,9 @@ constexpr bool within(Numeric value, Numeric floor, Numeric ceiling) {
 constexpr Numeric clamp(Numeric value, Numeric floor, Numeric ceiling) {
+  if(floor > ceiling){
+    floor = ceiling;
+  }
```
#380 specifies that if floor is greater than ceiling, we should swap floor and ceiling (which you can do using e.g. `std::swap`). This implementation instead just sets the floor to the ceiling, which is not intuitive behavior.

ethanjli reviewed (commented):
Here's my review of the changes to `Util` function implementations for #380 and #379.

rohanpurohit commented:
It looks like we'll be moving this to v0.13 and finish it as early as possible @ethanjli

raavilagoo committed:
Improved Structure, cpp files and txt files for following files
1.Array
2.Bytes
3.Ranges
4.Vector

raavilagoo committed:
Unit Test for Util
Added more tests in Array
Ranges - refined
Vectors - Improved structure
Byte - Improved txt file

raavilagoo committed:
Test cases
1.Vector

raavilagoo committed:
Modified Structure and test cases for
1. Array
2. Bytes
3. Vector
4. Ranges

raavilagoo commented:
@ethanjli ,This pull request is ready for review.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Bytes.txt:
```diff
@@ -0,0 +1,74 @@
+     THEN('the function returns the value shifted left into 0th position, which is 0x86')
+
+  WHEN('the set_byte function   is called with byte index 1 and output type uint16_t as the template parameters and with 0x83 as the input value')
+     THEN('the function returns the value shifted left into 1st position, which is 0x8600')
```
[proofreading] The input and output values are inconsistent.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,38 @@
+
+Scenario: Util:Array make_array function works correctly for uint8, uint16, uint32, and int16 types
```
[coverage & traceability] It would be better if you also tested a `struct` here for basic functionality. The next scenario tests structs, but it also tests `std::reference_wrapper`s, i.e. two things are changed together and we don't have a test where we only change one thing yet. I think it's reasonable to expect that if `make_array` works on one primitive type (such as a `uint`) then it will work on all of them, especially since `make_array` is unrelated to the specific bytes inside the type. It is not necessarily obvious that `make_array` works on `struct`s if we only have tests showing it works on ints. However, I will not block the PR on this if you don't want to do this.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,38 @@
+    WHEN('an array is created from two elements of structure having values 0x02 and 0x0f')
+      THEN(' the array has the expected values, 0x02 and 0x0f') 
+      THEN('Element values of structure are changed to 0x03 and 0xde, values of array at index 0 and at index 1 also change to 0x03 and 0xde') 
+      THEN('Values of an array can't be changed as elements in array are const reference')
```
[implementability of test] If we don't have a way to write compilable test code, maybe we should comment out this THEN and add a comment above it (in this file and in the cpp file) explaining that the code trying to change the values of the array doesn't compile. I think it's also fine to just omit this THEN entirely.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,38 @@
+  GIVEN('Util:Array make_array function') 
+    WHEN('an array is created from two elements of structure having values 0x02 and 0x0f')
+      THEN(' the array has the expected values, 0x02 and 0x0f') 
+      THEN('Element values of structure are changed to 0x03 and 0xde, values of array at index 0 and at index 1 also change to 0x03 and 0xde') 
```
[clarity] the wording should have a clearer cause-and-effect relationship. Right now it's ambiguous because you're just asserting two declarations ("Element values of structure are changed to 0x03 and 0xde" and also "values of array at index 0 and at index 1 also change to 0x03 and 0xde"): you could be asserting that when the array is created, it has the expected values and then it changes the elements of the structure on its own. Here's an alternative way you could write this:

```suggestion
      THEN('Changing the values in the structure to 0x03 and 0xde also changes values of array accessible at indices 0 and 1 to 0x03 and 0xde') 
```

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,38 @@
+
+Scenario: Util:Array make_array function works properly for const and non-const references
+  GIVEN('Util:Array make_array function') 
+    WHEN('an array is created from two elements of structure having values 0x02 and 0x0f')
```
[clarity] This should specify what kind of array we're making. We should be making an array of const references. Here's one way you could word this:

```suggestion
    WHEN('an array of const references is created from two struct instances with internal values of 0x67 and 0x3e') 
```

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,38 @@
+      THEN('Element values of structure are changed to 0x03 and 0xde, values of array at index 0 and at index 1 also change to 0x03 and 0xde') 
+      THEN('Values of an array can't be changed as elements in array are const reference')
+
+    WHEN('an array is created from two non const elements of structure with values 0x67 and 0x3e') 
```
[clarity] This should specify what kind of array we're making. We should be making an array of non-const references. The two structs themselves should not be const. Here's one way you could word this:

```suggestion
    WHEN('an array of non-const references is created from two struct instances with internal values of 0x67 and 0x3e') 
```

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,112 @@
+    WHEN('The clamp function is called with the function inputs') 
+      THEN('The clamp function returns the test value (8)') 
+
+    // floor > value == ceiling 
```
[proofreading] this should be changed to `floor < value == ceiling`. The tests for `floor > ceiling` belong in the third scenario, not the first scenario.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,112 @@
+
+SCENARIO: The Util::Ranges within and clamp function gives the expected result') 
```
[clarity] Let's call this scenario "The Util::Ranges within and clamp functions give the expected result for floor < ceiling"

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,112 @@
+      THEN('The Within function returns true') 
+
+    WHEN('The clamp function is called with the function inputs ') 
+      THEN('The clamp function returns the floor (15)') 
```
[proofreading] shouldn't this return the ceiling? Anyways, in your cpp file you used a test value of 15, which is inconsistent with this txt file. Also, your GIVENs/WHENs/THENs in the cpp file don't match the ones in this txt file. When I review tests I expect to be able to assume that the GIVENs/WHENs/THENs match between the files.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Ranges.txt:
```diff
@@ -0,0 +1,112 @@
+      THEN('The Within function returns false') 
+
+    WHEN('The clamp function is called with the function inputs') 
+      THEN('The clamp function returns the floor (8)')
```
[proofreading] 8 is actually the ceiling.

ethanjli reviewed (commented):
Here's my review of all the files besides vector.txt

raavilagoo committed:
Proofreading issues solved for
1. Array
2. Bytes
3. Ranges
4. Vector
In vector.h source_byte is renamed to source

raavilagoo commented:
@ethanjli ,This pull request is ready for review.

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,42 @@
+    WHEN('An array is created from input type int16_t as the template parameters and with 2 int16  input values') 
+      THEN('The array has the expected values of  int16_t at every index') 
+
+    WHEN('An array is created from two struct instances with internal values of 0xb9, 0xe2, 0xa7 and  0x31')
```
[proofreading] this should be four struct instances, right?

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt:
```diff
@@ -0,0 +1,42 @@
+    
+    WHEN('an array of non-const references is created from two struct instances with internal values of 0x67 and 0x3e')       
+      THEN('The array has the expected values, 0x67 and 0x3e') 
+      THEN('Values of non-const reference elements from structure are changed to 0xa3 and 0xee, values of array at index 0 and at index 1 changes to 0xa3 and 0xee') 
```
[clarity] the wording should have a clearer cause-and-effect relationship. Right now it's ambiguous because you're just asserting two declarations ("values ... are changed to ..." and also "values ... changes to ..."): you could be asserting that when the array is created, it has the expected values and then it changes the elements of the structure on its own.

raavilagoo committed:
Minor fixes in Array.txt file

raavilagoo commented:
@ethanjli , Pull request is ready to review.
I will make the expected changes in vector test cases in next pull request.

ethanjli commented:
Sounds good - there are other clarity issues which need to be improved in vector.txt: basically the same kinds of issues we discussed last meeting, except they're in the new changes you had added to vector.txt after that meeting. Since you plan to save the other changes for vector.txt for the next PR, let's merge this PR in for v0.13, and save those fixes for your next PR. For the next PR, let's try:

1. Making new commits for each round of changes you make, instead of force-pushing. I suspect this might be related to some of the weird behavior Github is doing with comment replies, and also force-pushing prevents me from seeing which changes you made between commits.
2. Before editing or creating any cpp files for test implementation, first we'll get the txt files for behavior specification through a few rounds of review so that they're mature.

For records-keeping: 

1. This project is licensed under Apache License v2.0 for any software, and Solderpad Hardware License v2.1 for any hardware - do you agree that your contributions to this project will be under these licenses, too?
2. Were any of these contributions also part of work you did for an employer or a client?
3. Does this work include, or is it based on, any third-party work which you did not create?

raavilagoo commented:
1.This project is licensed under Apache License v2.0 for any software, and Solderpad Hardware License v2.1 for any hardware - do you agree that your contributions to this project will be under these licenses, too? - Yes
2.Were any of these contributions also part of work you did for an employer or a client? -No
3.Does this work include, or is it based on, any third-party work which you did not create? No

raavilagoo commented:
@ethanjli , I have gone through your suggestions, 
1. I will avoid force pushing for future PRs.
2. I agree, It would be helpful if we can decide the scope of Unit tests, I can get started with txt files first. Then we can have discussion on same.
